homeassistant/components/weatherflow/__init__.py

```diff
@@ -0,0 +1,78 @@
+"""Get data from Smart Weather station via UDP."""
+from __future__ import annotations
+
+import logging
+
+from pyweatherflowudp.client import EVENT_DEVICE_DISCOVERED, WeatherFlowListener
+from pyweatherflowudp.device import EVENT_LOAD_COMPLETE, WeatherFlowDevice
+from pyweatherflowudp.errors import ListenerError
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.start import async_at_started
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [SENSOR_DOMAIN]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up WeatherFlow from a config entry."""
+
+    client = WeatherFlowListener(host=entry.data.get(CONF_HOST, "0.0.0.0"))
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    @callback
+    def device_discovered(device: WeatherFlowDevice) -> None:
+        _LOGGER.debug("Found a device: %s", device)
+
+        @callback
+        def add_device_if_started(hass: HomeAssistant, device: WeatherFlowDevice):
+            async_at_started(
+                hass,
+                lambda _: async_dispatcher_send(
```

Comment:
This needs to decorated with `@callback` or it will run in the executor

Reply:
It is decorated with a callback.. isnt it?
<img width="469" alt="image" src="https://github.com/home-assistant/core/assets/6491743/655a2c26-6aa6-434b-bc1d-09dd9bda9582">

Reply:
The lambda isn't

Reply:
```diff
diff --git a/homeassistant/components/weatherflow/__init__.py b/homeassistant/components/weatherflow/__init__.py
index 5f205a3ccb9..2a969d35f15 100644
--- a/homeassistant/components/weatherflow/__init__.py
+++ b/homeassistant/components/weatherflow/__init__.py
@@ -28,35 +28,43 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up WeatherFlow from a config entry."""
 
     client = WeatherFlowListener(host=entry.data.get(CONF_HOST, "0.0.0.0"))
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
 
     @callback
-    def device_discovered(device: WeatherFlowDevice) -> None:
+    def _async_device_discovered(device: WeatherFlowDevice) -> None:
         _LOGGER.debug("Found a device: %s", device)
 
         @callback
-        def add_device_if_started(hass: HomeAssistant, device: WeatherFlowDevice):
+        def _async_add_device_if_started(
+            hass: HomeAssistant, device: WeatherFlowDevice
+        ):
+            entry.async_on_unload(
                 async_at_started(
                     hass,
+                    callback(
                         lambda _: async_dispatcher_send(
-                    hass, f"{DOMAIN}_{entry.entry_id}_add_{SENSOR_DOMAIN}", device
+                            hass,
+                            f"{DOMAIN}_{entry.entry_id}_add_{SENSOR_DOMAIN}",
+                            device,
+                        )
                     ),
                 )
+            )
 
         entry.async_on_unload(
             device.on(
                 EVENT_LOAD_COMPLETE,
-                lambda _: add_device_if_started(hass, device),
+                lambda _: _async_add_device_if_started(hass, device),
             )
         )
 
-    entry.async_on_unload(client.on(EVENT_DEVICE_DISCOVERED, device_discovered))
+    entry.async_on_unload(client.on(EVENT_DEVICE_DISCOVERED, _async_device_discovered))
 
     try:
         await client.start_listening()
     except ListenerError as ex:
         raise ConfigEntryNotReady from ex
 
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     async def _async_handle_ha_shutdown(event: Event) -> None:
```

Reply:
Ah ok ... I removed the lambda - it wasn't needed any longer. Does that solve the issue?

Reply:
( just saw your Dif! :)